src/libs/Violations/ViolationsUtils.ts

```diff
@@ -0,0 +1,195 @@
+            }
+
+            // Remove 'categoryOutOfPolicy' violation if category is in policy
+            if (hasCategoryOutOfPolicyViolation && transaction.category && isCategoryInPolicy) {
+                newTransactionViolations = reject(newTransactionViolations, {name: 'categoryOutOfPolicy'});
+            }
+
+            // Remove 'missingCategory' violation if category is valid according to policy
+            if (hasMissingCategoryViolation && isCategoryInPolicy) {
+                newTransactionViolations = reject(newTransactionViolations, {name: 'missingCategory'});
+            }
+
+            // Add 'missingCategory' violation if category is required and not set
+            if (!hasMissingCategoryViolation && policyRequiresCategories && !transaction.category) {
+                newTransactionViolations.push({name: 'missingCategory', type: 'violation', userMessage: ''});
+            }
+        }
+
+        if (policyRequiresTags) {
+            const hasTagOutOfPolicyViolation = transactionViolations.some((violation) => violation.name === 'tagOutOfPolicy');
+            const hasMissingTagViolation = transactionViolations.some((violation) => violation.name === 'missingTag');
+            const isTagInPolicy = Boolean(policyTags[transaction.tag]?.enabled);
+
+            // Add 'tagOutOfPolicy' violation if tag is not in policy
+            if (!hasTagOutOfPolicyViolation && transaction.tag && !isTagInPolicy) {
+                newTransactionViolations.push({name: 'tagOutOfPolicy', type: 'violation', userMessage: ''});
+            }
+
+            // Remove 'tagOutOfPolicy' violation if tag is in policy
+            if (hasTagOutOfPolicyViolation && transaction.tag && isTagInPolicy) {
+                newTransactionViolations = reject(newTransactionViolations, {name: 'tagOutOfPolicy'});
+            }
+
+            // Remove 'missingTag' violation if tag is valid according to policy
+            if (hasMissingTagViolation && isTagInPolicy) {
+                newTransactionViolations = reject(newTransactionViolations, {name: 'missingTag'});
+            }
+
+            // Add 'missingTag violation' if tag is required and not set
+            if (!hasMissingTagViolation && !transaction.tag && policyRequiresTags) {
+                newTransactionViolations.push({name: 'missingTag', type: 'violation', userMessage: ''});
+            }
+        }
+
+        return {
+            onyxMethod: Onyx.METHOD.SET,
+            key: `${ONYXKEYS.COLLECTION.TRANSACTION_VIOLATIONS}${transaction.transactionID}`,
+            value: newTransactionViolations,
+        };
+    },
+    /**
+     * Gets the translated message for each violation type. Necessary because `translate` throws a type error if you
+     * attempt to pass it template strings, or members of a string literal union that contains both function and string
+     * members.
+     *
+     * The issue is with the type of {@link PhraseParameters} which is defined as
```

Comment:
Does this `{@link` thing make the editor render this doc nicely, and does it do the same for words wrapped in underscores and backticks? If not, should we remove that since our docs and comments are normally in plain text and not markdown?

Reply:
When you mouse over a function with JSDoc, the links function as actual links. So you can browse to the docs for those functions through the poppers. I'll provide some examples in the morning.

Reply:
![CleanShot 2023-12-12 at 12 38 35](https://github.com/Expensify/App/assets/22041394/db28a76b-1b25-4229-8751-f4a94b17c2b0)